### PR TITLE
Limit the timer resolution in non-isolated contexts

### LINKED ISSUE #89: Make high resolution time dependant on cross-origin isolated
With #88 solved, it would be good if the 5 microseconds resolution would be made conditional upon cross-origin isolated.

See https://github.com/whatwg/html/pull/4734. Though see https://github.com/whatwg/html/issues/5435 about potentially having to refer a slightly different concept.

### PR DESCRIPTION
**Pull request by yoavweiss**
Closes #89 

We still need to discuss and see what `X` value would be reasonable here

***
<a href="https://pr-preview.s3.amazonaws.com/yoavweiss/hr-time/pull/93.html" title="Last updated on Feb 18, 2021, 11:01 AM UTC (944206a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/hr-time/93/9b17e2b...yoavweiss:944206a.html" title="Last updated on Feb 18, 2021, 11:01 AM UTC (944206a)">Diff</a>

### CODE REVIEW

**Comment by npm1** [comment]
This makes sense but I think we want to move this to some algorithm which is called every time we want to get a timestamp, otherwise the `current settings object` does not make much sense.

**Comment by yoavweiss** [comment]
> This makes sense but I think we want to move this to some algorithm which is called every time we want to get a timestamp, otherwise the `current settings object` does not make much sense.

Good call. That would indeed be nicer and more precise.

**Comment by annevk** [comment]
require-corp does not guarantee isolation by itself. You want cross-origin isolated...

**Comment by yoavweiss** [comment]
> require-corp does not guarantee isolation by itself. You want cross-origin isolated...

Cool, thanks!

**Comment by annevk** [comment]
Thinking about this more you probably want cross-origin isolated capability here. So only if cross-origin isolated is delegated would a third party frame get true high-resolution timers.

cc @domenic @yutakahirano

**Comment by domenic** [comment]
Yes, we purposefully don't export "cross-origin isolated" since people should be using the capability.

**Comment by yoavweiss** [comment]
FYI - @mikewest @camillelamy @hemeryar

**Comment by yoavweiss** [comment]
Rebased the PR based on #97 and comments. Please take a look.

We still need to determine:
* What would a good value for `X` be?
* How can we publish this from a process perspective?
  - Do we need to go back to PR/CR? Define this as L3? Define HR-Time as a Living Standard as part of Process 2020?
* Do we need WPTs? If so, how can we actually test this?

**Comment by yutakahirano** [comment]
Could you use https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-cross-origin-isolated-capability instead of agent cluster's cross-origin isolated?

**Comment by yoavweiss** [comment]
> Could you use https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-cross-origin-isolated-capability instead of agent cluster's cross-origin isolated?

Exactly what I was looking for. Thanks!!

File: index.html [hunk 1]
```diff
@@ -262,9 +263,20 @@ <h3>Time Origin</h3>
...
-    |current global|.</p>
+    [=Realm/global object=] |current global| must run the following steps:
+    <ul>
+      <li>Let |current time| be the present time.</li>
+      <li>Let |settings object| be |current global|'s corresponding
+        [=environment settings object=].</li>
+      <li>Let |time resolution| be 5 microseconds.</li>
+      <li>If |settings object|'s
+        [=environment settings object/cross-origin isolated capability=] is set
+        to false, set |time resolution| to be X microseconds.</li>
```
Comment (annevk) [inline index.html:274]: is false*
what's X?

File: index.html [hunk 2]
```diff
@@ -262,9 +263,20 @@ <h3>Time Origin</h3>
...
+      <li>In a user-agent specific manner, jitter and coarsen |current time|
```
Comment (annevk) [inline index.html:275]: `<a>implementation-defined</a>`

File: index.html [hunk 3]
```diff
@@ -262,9 +263,20 @@ <h3>Time Origin</h3>
...
+        such that its resolution will not exceed |time resolution|.</li>
+      <li>Return the result of [=relative high resolution time=] where |time|
+        is |current time| and |global| is |current global|.</li>
```
Comment (annevk) [inline index.html:278]: given current time and current global* maybe? Using var for referencing named? variables and actual variables is rather confusing. (And would generate an error in Bikeshed.)

File: index.html [hunk 4]
```diff
@@ -380,13 +392,21 @@ <h3>Clock resolution</h3>
...
+      <p>To ensure that the new API does not significantly improve the
       accuracy or speed of such attacks, the minimum resolution of the
       {{DOMHighResTimeStamp}} type should be inaccurate enough to prevent
-      attacks: the current minimum recommended resolution is no less than 5
-      microseconds and, where necessary, should be set higher by the User
-      Agent to address privacy and security concerns due to architecture or
-      software constraints, or other considerations.</p>
+      attacks.</p>
+
+      <p>The current minimum recommended resolution is no less than 5
```
Comment (annevk) [inline index.html:401]: Hmm, I thought that's the resolution when there's no isolation? We try to return system time when you are isolated.

File: index.html (hunk 3 above)
Comment (yoavweiss) [reply index.html]: Changed!

File: index.html (hunk 2 above)
Comment (yoavweiss) [reply index.html]: done

File: index.html (hunk 1 above)
Comment (yoavweiss) [reply index.html]: Changed to "is false".

We still need to decide on the right value for `X`. I'm pushing for 100 microsec, as it seems to enable the relevant sub-ms use-cases without exposing too much granularity.

**Comment by yoavweiss** [comment]
There are 2 issues we still need to discuss it seems:
1) The value of the limit in the non-isolated context case
2) Removing the limit in the isolated context case

Regarding (1), I would go with 100 microseconds, as it's a value that seems to satisfy the use-cases without being overly granular.

The use-cases I found for sub-millisecond timers are:
* Ability to schedule work in sub-millisecond intervals
   - When running work on the main thread developers have a "window" of a few ms to do all the work they need to do for the browser to properly generate the next frame. 1ms granularity makes it hard to schedule such work without causing jank.
* RUM based benchmarking - getting accurate timing for e.g. User Timing can matter, depending on what you're measuring.
* Benchmarking (one [example](https://surma.dev/things/is-postmessage-slow/#:~:text=When%20Spectre%20%26%20Meltdown%20were%20discovered%2C%20all%20browsers%20disabled)) - that seems like a less critical example, as folks who want to benchmark specific browser operations can do so in isolated contexts
* [Framerate calculations and audio/animation coordination](https://developers.google.com/web/updates/2012/08/When-milliseconds-are-not-enough-performance-now#use_cases)

Regarding (2), I'm not convinced that isolated contexts resolve the [spy in the sandbox](http://www.cs.columbia.edu/~simha/spyjs.ccs15.pdf) caching attacks.
Talking to Chrome security folks, Chrome will not remove the limitation, even if the spec says it's no longer mandatory.
And while it's true that SAB and threads expose better granularity in isolated contexts, thread scheduling could mitigate that if that turns out to be a problem in practice.

At the same time, it may make sense to relax it in spec, as e.g. it may not be an issue on non-Intel hardware, and implementation-defined limitations could apply the limit only where it makes sense from a security perspective.

@annevk - Would you be interested in joining a WG call to discuss the above?

**Comment by annevk** [comment]
I chatted with Tom Ritter and my understanding about Firefox was incorrect. Firefox also does clamping:

1. Cross-origin isolated: clamped to 20 microseconds, no jitter.
2. Normal: clamped to 1000 microseconds, jitter.
3. Reduced fingerprinting preference enabled: clamped to 100000 microseconds, jitter (in both cross-origin isolated and normal).

We can change cross-origin isolated to clamp to 5 microseconds, but indeed, we want to keep clamping there as well and I was mistaken above.

**Comment by yoavweiss** [comment]
Thanks! I think we're in agreement, so no need to further discuss.
Note that the spec setting a 5 microsec minimum doesn't commit Firefox to change its clamping from e.g. 20 microsec.

**Comment by annevk** [comment]
It's not currently phrased as a minimum though. Should it be "an implementation-defined number of microseconds that is at least 5" or some such?

**Comment by yoavweiss** [comment]
> It's not currently phrased as a minimum though. Should it be "an implementation-defined number of microseconds that is at least 5" or some such?

I thought https://github.com/w3c/hr-time/pull/93/files#diff-eacf331f0ffc35d4b482f1d15a887d3bR275 covered the implementation-defined part where |time resolution| is a minimum. But re-reading that, I can make it more explicit.

**Comment by annevk** [comment]
I think instead of having that out-of-band, it would be preferable to have that flexibility directly stated where you introduce the constants, i.e., in the algorithm where you assign time resolution the value 5 microseconds. It's confusing to have normative requirements in multiple places, especially when they are in conflict if you read it literally.

**Comment by yoavweiss** [comment]
> I think instead of having that out-of-band, it would be preferable to have that flexibility directly stated where you introduce the constants, i.e., in the algorithm where you assign time resolution the value 5 microseconds.

I just changed the algorithm to do that.


> It's confusing to have normative requirements in multiple places, especially when they are in conflict if you read it literally.

Yeah, I can remove the explicit values from the privacy section, and point folks towards the processing model.

File: index.html (hunk 4 above)
Comment (yoavweiss) [reply index.html]: Resolved in https://github.com/w3c/hr-time/pull/93#issuecomment-704337546, I believe

File: index.html (hunk 1 above)
Comment (yoavweiss) [reply index.html]: I've set X to 100microsec, given no one objected.

**Comment by yoavweiss** [comment]
@annevk @npm1 - PTAL?

File: index.html [hunk 5]
```diff
@@ -262,9 +263,23 @@ <h3>Time Origin</h3>
...
+    <ul>
+      <li>Let |current time| be the present time.</li>
+      <li>Let |settings object| be |current global|'s corresponding
+        [=environment settings object=].</li>
+      <li>Let |time resolution| be 5 microseconds, or a higher
+        <a>implementation-defined</a> value.</li>
+      <li>If |settings object|'s
+        [=environment settings object/cross-origin isolated capability=] is
+        false, set |time resolution| to be 100 microseconds, or a higher
+        <a>implementation-defined</a> value.</li>
```
Comment (annevk) [inline index.html:276]: Can we flip these two steps, so we start high and go lower if the correct flag is set?

File: index.html [hunk 6]
```diff
@@ -262,9 +263,23 @@ <h3>Time Origin</h3>
     </div>
     </p>
     <p>The <dfn data-export="">current high resolution time</dfn> given a 
-    [=Realm/global object=] |current global| returns the result of <a>relative
-    high resolution time</a> where |time| is the present time and |global| is 
-    |current global|.</p>
+    [=Realm/global object=] |current global| must run the following steps:
```
Comment (annevk) [inline index.html:266]: I think you want `=/global object=` here, not a member of Realm that happens to contain one.

File: index.html [hunk 7]
```diff
@@ -380,13 +395,17 @@ <h3>Clock resolution</h3>
...
+      <p>To ensure that the new API does not significantly improve the
       accuracy or speed of such attacks, the minimum resolution of the
       {{DOMHighResTimeStamp}} type should be inaccurate enough to prevent
-      attacks: the current minimum recommended resolution is no less than 5
-      microseconds and, where necessary, should be set higher by the User
-      Agent to address privacy and security concerns due to architecture or
-      software constraints, or other considerations.</p>
+      attacks.</p>
+
+      <p>Where necessary, the User Agent should set higher resolution values to
```
Comment (annevk) [inline index.html:404]: user agent*

File: index.html [hunk 8]
```diff
@@ -262,9 +263,23 @@ <h3>Time Origin</h3>
...
     </p>
     <p>The <dfn data-export="">current high resolution time</dfn> given a 
-    [=Realm/global object=] |current global| returns the result of <a>relative
-    high resolution time</a> where |time| is the present time and |global| is 
-    |current global|.</p>
+    [=Realm/global object=] |current global| must run the following steps:
+    <ul>
+      <li>Let |current time| be the present time.</li>
+      <li>Let |settings object| be |current global|'s corresponding
+        [=environment settings object=].</li>
```
Comment (annevk) [inline index.html:270]:
```suggestion
      <li>Let |settings object| be |current global|'s [=relevant settings object=].</li>
```

**Review by annevk — COMMENTED** [review]
Found some nits, looks good overall though.

File: index.html (hunk 5 above)
Comment (yoavweiss) [reply index.html:276]: Yup

File: index.html (hunk 7 above)
Comment (yoavweiss) [reply index.html:404]: Yup

File: index.html (hunk 6 above)
Comment (yoavweiss) [reply index.html:266]: Done

File: index.html (hunk 8 above)
Comment (yoavweiss) [reply index.html:270]: Applied that change locally.

**Review by annevk — APPROVED** [review]
Have a white checkmark. Thanks Yoav!

File: index.html [hunk 9]
```diff
@@ -262,9 +263,22 @@ <h3>Time Origin</h3>
...
+        <a>implementation-defined</a> value.</li>
+      <li>If |settings object|'s
+        [=environment settings object/cross-origin isolated capability=] is
+        true, set |time resolution| to be 5 microseconds, or a higher
+        <a>implementation-defined</a> value.</li>
+      <li>In an <a>implementation-defined</a> manner, coarsen and potentially
+        jitter |current time| such that its resolution will not exceed |time
+        resolution|.</li>
+      <li>Return the result of [=relative high resolution time=] given
+        |current time| and |current global|.</li>
```
Comment (npm1) [inline index.html:280]: This method just returns the difference between the coarsened timestamp and the timeOrigin, is the next step to coarsen the value stored so this works as intended and the timestamp returned looks coarse as well? FWIW in Chrome that is what we're doing, although timeOrigin itself does not seem to return a coarsened value, but per your issue it probably should.

**Review by npm1 — APPROVED** [review]

File: index.html (hunk 9 above)
Comment (yoavweiss) [reply index.html:280]: Yeah. Alternatively, we could coarsen the output of the diff between the two. Happy to change that if you prefer.

Comment (npm1) [reply index.html:280]: I prefer it this way as it's aligned with Chrome impl, not sure about others though.

Comment (yoavweiss) [reply index.html]: Done!
@annevk - can you take another quick look?

Comment (annevk) [reply index.html]: I suspect you misunderstood @npm1 about how Chrome is doing it. As long as the end result is not black box distinguishable I don't really care.

Comment (npm1) [reply index.html]: Yea sorry for the confusion what I mean is that the implementation does this:
* Let `coarse current time` be the result of coarsening `current time`
* Let `coarse time origin` be the result of coarsening `time origin`
* Return `coarse current time` - `coarse time origin`

So no need to coarsen the difference in this PR. Instead we want a PR that will coarsen `time origin` (both for this computation and for `timeOrigin` return value)

Comment (yoavweiss) [reply index.html:280]: OK. I reverted my latest change, and we can coarsen the time origin in a separate PR.

**Pull request merged**